plotly/tools.py

```diff
@@ -1269,3 +1276,547 @@ def return_figure_from_figure_or_data(figure_or_data, validate_figure):
+
+    # Make arrays of x & y end values
+    Xend = x + (u * scale)
+    Xend = Xend.flatten('F')
+    Yend = y + (v * scale)
+    Yend = Yend.flatten('F')
+
+    # Make array of nans
+    # These are used as spaces so the separate
+    # arrows can be included in 1 trace
+    nnn = np.empty((len(Xstart)))
+    nnn[:] = np.NAN
+
+    # Combine arrays into matrix
+    Xvals = np.matrix([Xstart, Xend, nnn])
+    # Make matrix into array readable by plotly
+    Xvals = np.array(Xvals)
+    Xvals = Xvals.flatten('F')
+    Xvals = Xvals.tolist()
+
+    # Combine arrays into matrix
+    Yvals = np.matrix([Ystart, Yend, nnn])
+    # Make matrix into array readable by plotly
+    Yvals = np.array(Yvals)
+    Yvals = Yvals.flatten('F')
+    Yvals = Yvals.tolist()
+
+    # Make Trace1: the lines
+    trace1 = Scatter(x=Xvals, y=Yvals,
+                     mode='lines', name='Barb',
+                     line=Line(width=barb_width, color=barb_color))
+
+    # Arrows
+    # Get line lengths
+    # Default arrow length = 30% of line length
+    Xdif = Xend - Xstart
+    Ydif = Yend - Ystart
+    LineLen = np.sqrt(np.square(Xdif) + np.square(Ydif))
+    ArrowLen = LineLen * arrow_scale
+
+    # Get angle of line
+    LineAng = np.arctan(Ydif / Xdif)
+
+    # Set angles to create arrow
+    # Default angle = +/- 20 degrees
+    Ang1 = LineAng + (angle)
+    Ang2 = LineAng - (angle)
+
+    XSeg1 = np.cos(Ang1) * ArrowLen
+    YSeg1 = np.sin(Ang1) * ArrowLen
+
+    XSeg2 = np.cos(Ang2) * ArrowLen
+    YSeg2 = np.sin(Ang2) * ArrowLen
+
+    XPoint1 = np.empty((len(Xdif)))
+    YPoint1 = np.empty((len(Ydif)))
+
+    XPoint2 = np.empty((len(Xdif)))
+    YPoint2 = np.empty((len(Ydif)))
+
+    for index in range(len(Xdif)):
+        if Xdif[index] >= 0:
+            XPoint1[index] = Xend[index] - XSeg1[index]
+            YPoint1[index] = Yend[index] - YSeg1[index]
+            XPoint2[index] = Xend[index] - XSeg2[index]
+            YPoint2[index] = Yend[index] - YSeg2[index]
+        else:
+            XPoint1[index] = Xend[index] + XSeg1[index]
+            YPoint1[index] = Yend[index] + YSeg1[index]
+            XPoint2[index] = Xend[index] + XSeg2[index]
+            YPoint2[index] = Yend[index] + YSeg2[index]
+
+    # Combine arrays into matrix
+    XArrows = np.matrix([XPoint1, Xend, XPoint2, nnn])
+    # Make matrix into array readable by plotly
+    XArrows = np.array(XArrows)
+    XArrows = XArrows.flatten('F')
+    XArrows = XArrows.tolist()
+
+    # Combine arrays into matrix
+    YArrows = np.matrix([YPoint1, Yend, YPoint2, nnn])
+    # Make matrix into array readable by plotly
+    YArrows = np.array(YArrows)
+    YArrows = YArrows.flatten('F')
+    YArrows = YArrows.tolist()
+
+    # Make trace2: the arrows
+    trace2 = Scatter(x=XArrows, y=YArrows,
+                     mode='lines', name='Arrow',
+                     line=Line(width=arrow_width, color=arrow_color))
+
+    # Data
+    data = Data([trace1, trace2])
+
+    return data
```

Comment:
:cow2: you might consider making this into a class? this function has > 100 lines of code in it. As an example, it can then have a single method that handles cleaning up the input arguments, which currently takes up ~20 lines.